aten/src/ATen/functorch/TensorWrapper.h

```diff
@@ -89,7 +89,7 @@ struct TORCH_API TensorWrapper : public c10::TensorImpl {
   std::shared_ptr<bool> is_alive_;
 };
 
-TORCH_API Tensor makeTensorWrapper(const Tensor& tensor, int64_t level, bool generated=false);
+TORCH_API Tensor makeTensorWrapper(const Tensor& tensor, int64_t level, bool alias_of_unwrapped=false);
```

Comment:
I'm still thinking through this PR. Here's a brain dump of where I'm at:
- The wrap-call-unwrap we are doing in this PR. I see why it is necessary, but it makes me nervous. One way around it is that we say if we are doing a view operation and ALL tensors are not wrapped, then we ignore the current grad layer. This is what we do for vmap (if the tensors are not batched at the current level, then ignore the current vmap layer). If we are doing a view operation with some wrapped tensors and some unwrapped tensors, then we can error out.
- There's a BC-breaking vs silently incorrect behavior tradeoff. I don't want to accidentally make some behavior that people do rely on today error out.
- Performance. Checking alias annotations on schemas on every dispatch call might be a problem
- in-place view operations: do we have a problem with those?

Reply:
Another way around this is: we introduce some `is_immutable` flag to TensorWrapper. Tensors that get automatically wrapped in a TensorWrapper get the `is_immutable` flag. View operations propagate the `is_immutable`; out-of-place operations don't propagate it, and in-place operations cause an error if the TensorWrapper is immutable.

Pros: No more BC-breaking vs silently incorrect behavior tradeoff. Only silently incorrect behavior will be disallowed
Cons: Could be difficult to implement

Reply:
> Another way around this is: we introduce some is_immutable flag to TensorWrapper and keep the existing wrapping logic. Tensors that get automatically wrapped in a TensorWrapper get the is_immutable flag. View operations propagate the is_immutable; out-of-place operations don't propagate it, and in-place operations cause an error if the TensorWrapper is immutable.

I'm missing--how is that different than what's done here?

The previous name (generated) was a poor choice, so in case this was unclear: Originally I was thinking that we could just leave aliased tensors unwrapped. However, this caused some issues with the autograd meta. So, I ended up leaving the wrapping logic exactly as it was but if a tensor was unwrapped when it got to the front layer, the wrapper generated for it would have a flag on it (currently called `alias_of_unwrapped`) and that flag would only be propagated if the output is an alias of the input (based on the schema's alias info). If we tried to do run an inplace op on a tensor with this flag, we fail

> Performance. Checking alias annotations on schemas on every dispatch call might be a problem

Yeah this is a problem. I was thinking that we could memoize the function to start with? In the future, we might also be able to add this as a property that we compute when we build the schemas but I don't know the function schema code super well

> in-place view operations: do we have a problem with those?

To check, you're talking about ops marked with the `inplace_view` tag. For those, they match the [inplace op logic](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/functorch/DynamicLayer.cpp#L353) so they will have [always failed with an unwrapped input](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/functorch/ADInterpreters.cpp#L7)

Reply:
> I'm missing--how is that different than what's done here?

You're right -- that's exactly what this PR is doing! I was confused because I didn't read the PR, I was going off of what I thought the design was, which was that we were attempting to return an unwrapped Tensor if the aliased tensor was unwrapped

Reply:
> > in-place view operations: do we have a problem with those?
> 
> To check, you're talking about ops marked with the inplace_view tag. For those, they match the [inplace op logic](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/functorch/DynamicLayer.cpp#L353) so they will have [always failed with an unwrapped input](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/functorch/ADInterpreters.cpp#L7)

Yes, that's a good point.

So the problem right now is figuring out the performance